File: ListObjC/ListObjC/Factories/TableViewFactory.m
//
//  TableViewFactory.m
//  ListObjC
//
//  Created by pablo borquez on 26/08/23.
//

#import "TableViewFactory.h"
#import "TableViewCell.h"

@implementation TableViewFactory

- (void)prepareFactory:(UITableView *)tableView {
    [tableView setDataSource:self];
    [tableView setDelegate:self];
    
    UINib *nib = [UINib nibWithNibName:@"TableViewCell" bundle:nil];
    [tableView registerNib:nib forCellReuseIdentifier:@"TableViewCell"];
}

// MARK: - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 15;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    TableViewCell *cell = (TableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"TableViewCell"
                                                                           forIndexPath:indexPath];
  
    [[cell titleLabel]setText:[NSString stringWithFormat:@"Row: %@", @([indexPath row])]];
    
    return cell;
}

// MARK: - UITableViewDelegate

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 44;
}

@end

File: ListSwift/ListSwift/Factories/TableViewFactory.swift
//
//  TableViewFactory.swift
//  ListSwift
//
//  Created by pablo borquez on 08/09/23.
//

import UIKit

class TableViewFactory: NSObject, UITableViewDataSource, UITableViewDelegate {
    func prepare(tableView: UITableView) {
        tableView.dataSource = self
        tableView.delegate = self
        
        let nib = UINib(nibName: "TableViewCell", bundle: Bundle.main)
        tableView.register(nib, forCellReuseIdentifier: "cell")
    }
    
    // MARK: - UITableViewDataSource
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 15
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! TableViewCell
        cell.titleLabel.text = "Row: \(indexPath.row)"
        
        return cell
    }
    
    // MARK: - UITableViewDelegate
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 44
    }
}
